lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -2083,47 +2083,90 @@ void GrandCentralPass::runOnOperation() {
                   hw::OutputFileAttr::getFromFilename(
                       &getContext(), maybeExtractInfo->bindFilename.getValue(),
                       /*excludeFromFileList=*/true));
-              op->setAttr("output_file",
-                          hw::OutputFileAttr::getFromDirectoryAndFilename(
-                              &getContext(),
-                              maybeExtractInfo->directory.getValue(),
-                              op.getName() + ".sv",
-                              /*excludeFromFileList=*/true,
-                              /*includeReplicatedOps=*/true));
-              op->setAttr("comment",
-                          builder.getStringAttr("VCS coverage exclude_file"));
-
-              // Look for any blackboxes instantiated by the companion and mark
-              // them for inclusion in the Grand Central extraction directory.
-              SmallVector<FModuleOp> modules({op});
-              DenseSet<Operation *> bboxes, visited({op});
+
+              // Look for any modules/extmodules _only_ instantiated by the
+              // companion.  If these have no output file attribute, then mark
+              // them as being extracted into the Grand Central directory.
+              SmallVector<Operation *> modules({op});
+              DenseSet<Operation *> visited({op});
+              InstanceGraphNode *companionNode =
+                  instancePaths->instanceGraph.lookup(op);
+
+              LLVM_DEBUG({
+                llvm::dbgs() << "Found companion module: "
+                             << companionNode->getModule().moduleName() << "\n"
+                             << "  submodules exclusively instantiated "
+                                "(including companion):\n";
+              });
               while (!modules.empty()) {
```

Comment:
I think you might be able to use the `df_iterator` here, and then you wouldn't need to manage the `modules` or `visited` set
```
for (auto node : llvm::df_iterator(companionNode) {
  auto mod = node->getModule // or something
```

Reply:
Awesome suggestion! `llvm::depth_first` removes a lot of the complexity.